design/proposals/ksonnet-lib-simple-constructors.md

```diff
@@ -0,0 +1,109 @@
+* ksonnet-lib should be a direct translation of the OpenAPI swagger.json.
+
+## Goal
+
+* Developers should be able to construct objects using the builder pattern rather than constructors.
+* Upstream projects should have a clean slate to build abstraction upon.
+
+## Non Goals
+
+The following are related but not the goals of this specific proposal
+
+* Create upstream projects that implement constructors (e.g. v1 Deployment)
+
+## Proposal
+
+The change would be introducing a new constructor for all resources.
+
+```jsonnet
+group:: {
+  v1:: {
+    local apiVersion = {apiVersion: "group/v1"},
+    kind:: {
+      local kind = {kind: "Kind"},
+      init() :: apiVersion + kind,
+    },
+  },
+}
+```
+
+Many of the existing constructors adhere to this format. This change would make sure all resources have a constructor that only includes the apiVersion and kind.
+
+## User Experience
+
+### Use Cases
+
+To deprecate the existing the existing custom constructors a new constructor will be introduced, `init()`. Currently, to construct a v1beta2 Deployment, you have to do the following:
+
+```jsonnet
+local params = ...;
+local deployment = k.apps.v1beta2.deployment;
+local container = k.apps.v1beta2.deployment.mixin.spec.template.spec.containersType;
+local containerPort = container.portsType;
+local labels = {app: params.name};
+local targetPort = params.containerPort;
+
+local appDeployment = deployment
+  .new(
+    params.name,
+    params.replicas,
+    container
+      .new(params.name, params.image)
+      .withPorts(containerPort.new(targetPort)),
+    labels
+);
+```
+
+This snippet illustrates a few issues:
+
+* The constructor requires items that aren't required by the spec.
+* The constructor requires a specific order.
+
+Instead of the custom constructor, instead you can construct using init using the following:
+
+```jsonnet
+local params = ...;
+local deployment = k.apps.v1beta2.deployment;
+local containersType = k.apps.v1beta2.deployment.mixin.spec.template.spec.containersType;
+local containerPort = container.portsType;
+local labels = {app: params.name};
+local targetPort = params.containerPort;
+
+local port = containerPort
+  .init()
+  .withContainerPort(targetPort)
+
+local container = containersType
+  .init()
+  .withName(params.name)
+  .withImage(params.image)
+  .withPorts(port)
+
+local appDeployment = deployment
+  .init()
+  .mixin.metadata.withName(params.name)
+  .mixin.spec.withReplicas(params.replicas)
+  .mixin.spec.template.spec.withContainers(containers)
+  .mixin.spec.template.metadata.withLabels(labels);
+```
+
+### Backwards compatibility
+
+Existing constructors will not be removed. The inclusion of `init()` constructors. Will be the only change. This change will not impact any of the current usages of ksonnet-lib. After a period of time (at most two releases of ksonnet), the existing constructors will be removed.
```

Comment:
Rewording for clarity:

This change will not immediately impact any of the current usages of ksonnet-lib. To begin, the inclusion of `init()` constructors will be the only change. Only after a period of time (at most two releases of ksonnet) will existing constructors be removed.

Reply:
if we end up doing `ks lint` we can probably add a warning when we actually deprecate the constructors and the user's ksonnet still has them